In response to bug 763640

Bob-IT opened the pull request:
Came across this bug raised by Geert and had a go at fixing. All works for me and tested on windows so hope fully nothing too drastic too change. If it looks OK, will make the changes on remaining scheme files that define a plot size.

Bob

gjanssens commented on src/html/gnc-html-webkit.c:
```diff
@@ -108,7 +108,7 @@ static void impl_webkit_show_url( GncHtml* self, URLType type,
                                   gboolean new_window_hint );
 static void impl_webkit_show_data( GncHtml* self, const gchar* data, int datalen );
-static void impl_webkit_reload( GncHtml* self );
+static void impl_webkit_reload( GncHtml* self, gboolean view );
```
"view" as variable name here is not really describing very well what the boolean is doing. Better would have been for example "force_reconstruct" or  "force_reparse" or something like that although those names invert the meaning of the boolean. But with a name like that it's easier to guess what it does.

gjanssens commented on src/app-utils/options.scm:
```diff
@@ -1143,6 +1143,49 @@
 
+;; number range options use the option-data as a list whose
+;; elements are: (lower-bound upper-bound num-decimals step-size)
+;; positve value numbers are pixels, negative numbers are percentages
```
Hmm, I'm not too fond of this construct. While technically this is correct it's not a direct mapping of the actual information (absolute value vs relative value) and hence takes a mental conversion each time.
I'd prefer a solution similar to the date widget which also has two ways to pass a date (absolute or relative). It stores this information as a pair.
That approach would work well here too imo and would be much clearer to follow.

gjanssens commented on src/gnome-utils/dialog-options.c:
```diff
@@ -2420,6 +2420,149 @@ gnc_option_set_ui_widget_dateformat (GNCOption *option, GtkBox *page_box,
+                     G_CALLBACK(gnc_option_changed_widget_cb), option);
+
+    px_butt = gtk_radio_button_new_with_label(NULL, _("Pixels"));
+    gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON(px_butt), TRUE);
```
Heh, while that is how it has always been, I think we want the percent option to be the default. I think a full width/height chart is more often useful than that 400px by 400px. The latter is almost always too small.

gjanssens commented on src/report/report-system/html-barchart.scm:
```diff
@@ -37,6 +37,8 @@
                       row-labels-rotated?
 		      stacked?
+                      pixels-width?
+                      pixels-height?
```
Building further on the idea of using a pair to store the width and height information (as in "(absolute . 400)" or "(relative . 100)"), there is no need to add these two extra parameters. Instead, the information you are looking for is directly encoded in the width and height parameters passed to this module.

gjanssens commented on src/report/report-system/html-barchart.scm:
```diff
@@ -58,7 +60,7 @@
 (define (gnc:make-html-barchart)
   (gnc:make-html-barchart-internal -1 -1 #f #f #f #f '() '() '() 
-				   #f #f #f '() #f #f #f #f #f #f))
+				   #f #f #f #t #t '() #f #f #f #f #f #f))
```
It's the first two that parameters here that should change to '(relative . -1) to be initialized as a pair.

gjanssens commented on src/report/report-system/html-barchart.scm:
```diff
@@ -392,9 +406,15 @@
-            (push "\" style=\"width: 100%;height:100%;\"></div>\n")
+            (push "<div id=\"")(push chart-id)(push "\" style=\"width:")
+            (push (gnc:html-barchart-width barchart))
+            (if (gnc:html-barchart-pixels-width? barchart)
```
Check for (car (gnc:html-barchart-width width)) to be relative or absolute

gjanssens commented on src/report/report-system/html-linechart.scm:
```diff
@@ -38,6 +38,8 @@
                       row-labels-rotated?
                       stacked?
+                      pixels-width?
+                      pixels-height?
```
Same here of course...

gjanssens commented on src/report/standard-reports/category-barchart.scm:
```diff
@@ -203,13 +204,15 @@ developing over time"))
+    (gnc:options-add-plot-sizes! 
      options gnc:pagename-display 
-     optname-plot-width optname-plot-height "e" 400 400)
+     optname-plot-width optname-plot-height "f" 400 400)
```
Make the relative plot size the default here as well.

gjanssens commented:
Bob, thanks for working on this. I built with your patch and I like the new behaviour very much. Using the idle function is good research.

Apart from the inline comments I also noted that when multiple reports are open, only the visible one will react to window size changes. So after resizing and then switching back to another report, it will not have resized to fit the new window. Maybe you have to add a listener in report pages to listen for screen resizes or maybe the expose function was needed. I haven't investigated this further.

Bob-IT commented:
OK, I will investigate further your suggestions, have trouble understanding the scheme but will have another look at the date widget for inspiration.

Bob-IT commented:
Have made changes to variable view and fixed the page size updates by tracking the view size and then doing the update on an expose event. Also changed the plot size option to save a pair as suggested and after much counting of brackets this works. The only thing I can not find a way to test is the kvp->scm and scm->kvp procedures, are they used or can you point me in the right direction ?

The new plot size will read the old pixel sizes but will not update the saved report till it is saved again.

Regarding old version opening new one, this may be possible by changing the make-number-range-option used by the existing plot size options but that would effect all spinners, not sure.

Bob-IT commented:
Did not realise there were tests for some reports, these tests failed test-standard-net-barchart and test-standard-net-linechart, will run locally and fix.

Bob-IT commented:
OK, I think I have finished the report changes. All tests pass this time so hope fully looking good. I think I have opened all the reports and all ran successfully using guile 1.8.x. I will look at the fancy date changes in a separate request along with the default date format checks/changes.

jralls commented:
Being a bug fix, shouldn't this be on maint rather than master?

Bob-IT commented:
I did not think of this as a bug fix but more an enhancement as it would not show up on maint.
If the consensus think it should be against maint, then I will scrap this and apply the relevant patches against maint.

gjanssens commented:
Apart from one important detail this could have gone into maint as far as I'm concerned. That detail is this change breaks backwards compatibility in saved reports (and by extension reports still open while gnucash is closed). Gnucash 2.6.14 will choke on certain saved reports which are generated by the gnucash after this branch is merged.
I believe for that reason alone we can't apply this PR on maint as a whole. And this reminds we we need a follow-up PR on maint which ensures future 2.6.x version will be able to at least read the new format. And then the code should probably ignore it and display the report in a default fixed size anyway. The tricky part is in preserving what was saved anyway if *any other report gets saved*. That code will automatically resave each report and on 2.6.x that would mean losing the 'percent option. So perhaps we will also need to apply the internal code that can fully cope with the new sizing options and only ignore it for display and selection in the gui.

Bob-IT commented:
This is one aspect that I never really got my head round, why allow older version of Gnucash to read new version files ? Will need to think about what you have said.

gjanssens commented:
This is to allow users to fall back to the previous major version of gnucash if they can't fully migrate to the new version just yet. There can be many reasons for this. For example:
- a users runs gnucash on two different computers and can only upgrade gnucash on one of them for various reasons
- someone upgraded and only after a while a bug is found that requires him/her to revert to an older version until the bug is fixed

I realize most applications don't do this and I don't know when or why this policy was introduced in the gnucash project. It was there before I started using it.

gjanssens commented:
Bob, with the maint part out of the way, we're almost ready to merge in the master part as well. Since my last comment I remembered the saved-reports file is versioned. When we bump the version, gnucash will copy the old format to the new the first time the new version is queried. Older versions of gnucash continue to see the old version of the file.

As your code here changes the save format of some reports, I think it's worth increasing the saved-reports version so if a user saves the new percent format style it will still be there the next time the new gnucash version is run, regardless of what is done in the old gnucash version in between.

As I understand it, it suffices to change the version numbers in src/report/report-system/gnc-report.h
And as a cosmetic also change the two strings in src/report/report-gnome/gnc-plugin-page-report.c:
"The report will be saved in the file ~/.gnucash/saved-reports-2.4. "

jralls commented:
Concur. The new version should be '2.8'.
Ideally the strings in gnc-plugin-page-report.c should be composed using the version number from gnc-report.h so that we don't need to change them manually.

gjanssens commented:
Yes the string should be composed. I wasn't sure however how to manage this in relation to the translation of the string. The const str is surrounded by N_( ), which means to only mark this string as translatable, but don't translate it on the spot. Ideally the string marked for translation would be the generic string without the version number. But we don't control the code that actually does use the translated string, the code where the version should be inserted then. This code is part of the GtkActionEntry.

I suppose it's still better to compose the string here as you say. That means we won't forget to do so in future updates. And translators will get a fuzzy string, which is still better than no translation update because we forgot to update the version number. If you know a better way I'd love to hear it...

Bob-IT commented:
Do I need to wait for a merge from maint to master and then do a rebase ?

gjanssens commented:
Bob, that would probably be best. I meant to do this today already, but didn't get around to it. I'll try doing the merge tomorrow.

jralls commented:
> And translators will get a fuzzy string, which is still better than no translation update because we forgot to update the version number. If you know a better way I'd love to hear it...

The alternative would be to reword the phrase enough that it makes a new string instead of a fuzzy one. I guess that has an advantage that gettext won't be able to match on it and still report the wrong file name on translations that aren't up to date.

gjanssens commented:
No, that doesn't make any difference as I understand it. Gettext will ignore fuzzy translations as well. So in both cases the user will see the untranslated string with the correct version. "Fuzzy" is only an indication for translators, not for the end application.

When I asked for a better way I was hoping for one that would avoid the need for a translation update completely. But I don't see such a solution.

jralls commented:
Changing the msgid to "The report will be saved in the file %s" means only one more translation update. Had it been done right the first time then there wouldn't be one now. The substitution should be with the result of gnc_build_dotgnucash_path(SAVED_REPORTS_FILE).

BTW, currently we load both saved-reports-2.4 and saved-reports-2.0 (in gnucash-bin.c). With the bump should we load all three or drop 2.0?

gjanssens commented:
@Bob-IT I have merged maint into master just minutes ago. So you can start your rebase.

gjanssens commented:
@jralls I'll run a couple of experiments for the translation issue. I think we can hack around it indeed.
As for support for older saved-reports files, I propose dropping 2.0. The effect would be that users who last made changes to their saved reports in a gnucash version earlier than 2.2 will lose their saved reports when upgrading to gnucash 2.8. Perhaps we should add a note in the release notes regarding this. Something along the lines of:

> GnuCash 2.8 drops support for the saved report configurations format from gnucash 2.2 and earlier. If you have saved report configuration created in those versions of gnucash and never saved any report configuration later on in gnucash 2.4 or 2.6, you will find gnucash 2.8 no longer lists your saved report configurations. To avoid this, please save at least one report configuration in gnucash 2.4 or 2.6. You may delete it afterwards. 

This version bump would probably also allow us to drop some backwards compat code in the saved-reports area. I seem to remember the earlier format didn't save the report GUID and there's code in the report system to cater for that.

gjanssens commented:
I've pushed a fix for the translation issue. I think it's slightly hacky because I'm now passing a translated string to the GtkActionEntry instead of a string to translate. So GtkActionEntry will try to find a translation for the already translated string. I assume this will fail as we're not providing that and so the original translation as passed in will be used instead.
I don't see another way as I can't pass a parameterized untranslated string to the GtkActionEntry so this will have to to...

Bob-IT commented:
I was unable to get the rebase to work for me so see pull 112 for all the follow up updates after the maint merge.
Bob

Bob-IT commented:
Closing as updates applied in pull 112
